collect_stata/write_json.py

```diff
@@ -0,0 +1,369 @@
+            if var_value in stata_missings:
+                values.append(stata_missings[var_value])
+            else:
+                values.append(var_value)
+
+    cat_dict = sorting_dataframe(values, labels, missings, frequencies)
+
+    return cat_dict
+
+
+def uni_string():
+    """Generate dict with frequencies for nominal variables
+
+    Output:
+    string_dict: OrderedDict
+    """
+
+    string_dict = OrderedDict()
+
+    string_dict["frequencies"] = []
+    string_dict["labels"] = []
+    string_dict["missings"] = []
+    string_dict["values"] = []
+
+    return string_dict
+
+
+def uni_number():
+    """Generate dict with frequencies for numerical variables
+
+    Output:
+    number_dict: OrderedDict
+    """
+    number_dict = OrderedDict()
+
+    number_dict["frequencies"] = []
+    number_dict["labels"] = []
+    number_dict["missings"] = []
+    number_dict["values"] = []
+    number_dict["labels_de"] = []
+
+    return number_dict
+
+
+def stats_cat(elem, file_csv):
+    """Generate dict with statistics for categorical variables
+
+    Input:
+    elem: dict
+    file_csv: pandas DataFrame
+
+    Output:
+    statistics: OrderedDict
+    """
+
+    names = ["valid", "invalid"]
+    values = []
+
+    total = int(file_csv[elem["name"]].size)
+    invalid = int(file_csv[elem["name"]].isnull().sum()) + int(
+        sum(n < 0 for n in file_csv[elem["name"]])
+    )
+    valid = total - invalid
+
+    value_names = [valid, invalid]
+
+    for value_name in value_names:
+        values.append(str(value_name))
+
+    statistics = OrderedDict([("names", names), ("values", values)])
+
+    return statistics
+
+
+def stats_string(elem, file_csv):
+    """Generate dict with statistics for nominal variables
+
+    Input:
+    elem: dict
+    file_csv: pandas DataFrame
+
+    Output:
+    statistics: OrderedDict
+    """
+
+    names = ["valid", "invalid"]
+    values = []
+
+    int(file_csv[elem["name"]].size)
+    valid = int(file_csv[elem["name"]].value_counts().sum())
+    invalid = int(file_csv[elem["name"]].isnull().sum())
+    for i in file_csv[elem["name"]]:
+        if i in ("", "."):
+            valid = valid - 1
+            invalid = invalid + 1
+
+    value_names = [valid, invalid]
+
+    for value_name in value_names:
+        values.append(str(value_name))
+
+    statistics = OrderedDict([("names", names), ("values", values)])
+
+    return statistics
+
+
+def stats_number(elem, file_csv):
+    """Generate dict with statistics for numerical variables
+
+    Input:
+    elem: dict
+    file_csv: pandas DataFrame
+
+    Output:
+    statistics: OrderedDict
+    """
+
+    data_wm = file_csv[file_csv[elem["name"]] >= 0][elem["name"]]
+
+    names = ["Min.", "1st Qu.", "Median", "Mean", "3rd Qu.", "Max.", "valid", "invalid"]
+    values = []
+
+    mid = int(len(sorted(data_wm)) / 2)
+    first_q = np.median(sorted(data_wm)[:mid])
+    if len(sorted(data_wm)) % 2 == 0:
+        third_q = np.median(sorted(data_wm)[mid:])
```

Comment:
np.median() does not need a sorted list / array.
https://stackoverflow.com/a/16872271
https://docs.scipy.org/doc/numpy/reference/generated/numpy.median.html#numpy-median

Reply:
https://docs.scipy.org/doc/numpy/reference/generated/numpy.quantile.html

Reply:
You can simply use this line for both quartiles:

```python
first_q, third_q = np.quantile(data_wm, [0.25, 0.75])
```
I think you might even get the median in that call?
```python
first_q, median, third_q = np.quantile(data_wm, [0.25, 0.5, 0.75])
```
pandas wraps this for you with [describe](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.describe.html):

```python
>>> data_wm.describe()
count    11.000000
mean      1.090909
std       1.375103
min       0.000000
25%       0.000000
50%       1.000000
75%       1.500000
max       4.000000
Name: HKIND_Dummy, dtype: float64
```